Fix #2966: modified the the redirection rules when the user goes to root url

thedeveloperr opened the pull request:
I have implemented this feature as suggested by @jaredsilver . I modified HomePageRedirectHandler get() method. I retrieved the user contributions using user_services.get_user_contributions() function and checked the length of the list of contribution ids. if the length is more than 0 then user has explorations associated so my code redirect them to dashboard else redirect them to library page. I have Tested it locally, it's working as intended.

seanlip commented on main.py:
```diff
@@ -71,7 +71,16 @@ class HomePageRedirectHandler(base.BaseHandler):
     def get(self):
         if self.user_id and user_services.has_fully_registered(self.user_id):
-            self.redirect(feconf.DASHBOARD_URL)
+            # Get all the explorations associated with the user's account
```
The pattern seems reasonable, but please write backend tests for this in order to verify the behaviour. You can do so in base_test.py.

seanlip replied:
In fact, you should probably look at the Travis checks. They're failing, and this PR needs to address that too.

You may also want to take a look at the wiki documentation for how to run tests locally.

thedeveloperr replied:
```

      users.login('eve@privileges.com');
      general.openEditor(explorationId);
      // Eve is redirected to the homepage.
      expect(browser.getCurrentUrl()).toEqual(
        general.SERVER_URL_PREFIX + '/dashboard');
      users.logout();
    });
  });

```

this snippet is causing the fail. It assumes that homepage  is at /dashboard, I guess. But this feature request contradicts this assumption. So please guide me how shall I address this? As this is a test I am not sure if it should be modified or not.

seanlip replied:
You can modify tests appropriately in order to implement new features; just make sure that the modification makes sense.

Tests are meant to check that the behaviour matches the expected one. If the expected behaviour changes, it is logical for the tests to change accordingly to reflect the new behaviour.

thedeveloperr replied:
Shall I include the change in the test in same PR ?

seanlip replied:
Yes, please. In general, code should be accompanied by tests, and in any case you won't be able to submit the PR until all the tests have passed.

thedeveloperr replied:
> write backend tests for this in order to verify the behaviour. You can do so in base_test.py.

@seanlip  any links, documentation on how to do it. I am a newbie in these kind of stuff. I can learn fast if I can find a starting point on how to do it.

seanlip replied:
Have you tried reading the backend test files? They should be pretty self-explanatory.

thedeveloperr replied:
@seanlip sorry I opened different base_test.py file. I have read it and started to work. But while working on it there is a case where user once creates an exploration but then deletes it so that it has no exploration. How should it be treated ? Shall the user be redirected to library or dashboard?

jaredsilver replied:
If a user deletes an exploration, do we still associate that user with the deleted exploration in the database? If so, then we can still send that user to the creator dashboard because it is likely they are a creator.

thedeveloperr replied:
@jaredsilver yes it does, I manually created an exploration but then deleted it, it still shows that deleted contribution in the user contribution list ids 
`user_services.get_user_contributions(self.user_id).created_exploration_ids`
though it doesn't show that id's contribution in dashboard

jaredsilver replied:
Great! Then let's go with sending the user to the creator dashboard in that situation. Thanks, @thedeveloperr!

anmolshkl commented:
Hi @thedeveloperr, any update on the PR?

thedeveloperr commented:
@anmolshkl Made some commits, please check. 
1. Added some tests to automate testing of this feature
2.  modified existing tests to comply with this new feature

thedeveloperr commented:
How do I refresh a travis build ? I can't find refresh button as mentioned in wiki.

DubeySandeep commented:
Click on "Details" there you can find a refresh button :)

anmolshkl commented:
@thedeveloperr you need click on - show all checks -> details -> login to travis using github -> click on "Restart build" in top right corner or a small restart button next to the build job to restart a specific job only.

seanlip commented on core/controllers/base_test.py:
```diff
@@ -42,10 +42,23 @@
         super(BaseHandlerTest, self).setUp()
         self.signup('user@example.com', 'user')
 
+        # create a learner user to test learner redirect behavior
```
Please pay attention to punctuation, capitalization, etc., and make the comments crisp (e.g. what does "associated" mean)?

Create a learner user to test redirect behavior for the learner. A "learner" is defined as a user who has not created any explorations.

(Is this 'created', 'contributed to' or 'edited', btw? Be precise.)

thedeveloperr replied:
> Please pay attention to punctuation, capitalization, etc., 
Should I pay attention to punctuation ,capitalization etc. in comments or in 45-48 ? If in line 45-48 what specs should I follow?

seanlip replied:
In comments.

thedeveloperr replied:
Done

seanlip commented on core/controllers/base_test.py:
```diff
@@ -42,10 +42,23 @@
+        # learner is a user with no explorations associated with it
+        self.signup(self.TEST_LEARNER_EMAIL, self.TEST_LEARNER_USERNAME)
+
+        # create a creator user to test creator redirect behavior
```
Ditto.

thedeveloperr replied:
Done

seanlip commented on core/controllers/base_test.py:
```diff
@@ -107,9 +120,55 @@ def test_redirect_in_both_logged_in_and_logged_out_states(self):
+
+        # Homepage is either dashboard or library based on user's learner
+        # status, this homepage redirect rule behavior is seperately tested
+        atHome = ('dashboard' in response.headers['location'] or
```
Nope, test each case separately. There should be no uncertainty in the expected values for tests.

You can have three separate tests:
- logged out user
- logged in learner
- logged in creator

thedeveloperr replied:
@seanlip  @jaredsilver 
A new use case came to my mind. According to current approach users who have created an exploration i past are creators. But what if a user just edited another user's created explorations and not created any exploration ? Will that user be treated as a creator and be redirected to dashboard ?

jaredsilver replied:
@thedeveloperr Yes, I would say that if a user has edited another user's explorations then we should consider that user a creator and redirect him or her to the creator dashboard.

@seanlip Do we have any evidence that a learner would ever attempt to edit an exploration? If so, then perhaps I should reconsider my thoughts. Then again, I suspect this will get looked at again when we implement the learner dashboard, and doing this for now is probably better than not doing it.

seanlip replied:
@jaredsilver To your last point, I'd say no, not really. At most, they can give suggestions, but the eventual "author" for those is the person who actually merges it into the exploration. So what you propose in the first paragraph is fine.

thedeveloperr replied:
So basically should I move ahead and implement the new case of considering an editor a creator? @seanlip @jaredsilver

seanlip replied:
Yep, go ahead!

thedeveloperr replied:
Done

seanlip commented on core/controllers/base_test.py:
```diff
@@ -107,9 +120,55 @@ def test_redirect_in_both_logged_in_and_logged_out_states(self):
+        self.logout()
+
+    def test_root_redirect_rules_for_learners(self):
+        # Test learner behavior. Login as learner
```
I don't think you need this comment. It's pretty clear from both the method name and the test name.

thedeveloperr replied:
Done

seanlip commented on core/controllers/base_test.py:
```diff
@@ -107,9 +120,55 @@ def test_redirect_in_both_logged_in_and_logged_out_states(self):
+        # Test learner behavior. Login as learner
+        self.login(self.TEST_LEARNER_EMAIL)
+
+        # Dont create any exploration and simply go to root
```
`# Since no explorations have been created, going to '/' should redirect to the library page.`

thedeveloperr replied:
Done

seanlip commented on core/controllers/base_test.py:
```diff
@@ -107,9 +120,55 @@ def test_redirect_in_both_logged_in_and_logged_out_states(self):
+            exploration_id, creator_user_id, title='Test',
+            category='Test', language_code='en')
+
+        # go to root url
```
Here and elsewhere: for actions that are obvious from the code, you don't need a comment to describe what is going on. Use comments to explain _why_ you're doing things if needed, but don't use them if they just repeat what the code is obviously doing.

thedeveloperr replied:
Done

seanlip commented on core/tests/protractor.conf.js:
```diff
@@ -56,6 +56,7 @@ exports.config = {
   // When run without a command line parameter, all suites will run. If run
   // with --suite=smoke, only the patterns matched by that suite will run.
   suites: {
+
```
Don't modify this file.

thedeveloperr replied:
I did edit this file but I didn't added this file for commit my local repo, but this still got changed ! Does this mean if we edit file but dont commit or add this to staging area it will still be uploaded?

seanlip replied:
I don't know what you did, but you can fix it by modifying it back to what it used to be and adding another commit to this PR. After you push a PR, take a look at the "Files Changed" tab in the PR on GitHub to make sure that the diff is exactly what you want it to be.

thedeveloperr replied:
Done

seanlip commented on core/tests/protractor/privileges.js:
```diff
@@ -46,10 +46,18 @@ describe('Permissions for private explorations', function() {
+
+        // Homepage can be either dashboard or library dependin on learner
+        // Status
+        var atHomeFlag = (curUrl === libraryURL) || (curUrl === dashboardURL);
```
Nope. This test should be deterministic.

thedeveloperr replied:
In this test, from the source code, I know eve didn't contribute to any exploration so this means that I should check if the eve is on library page ? Or shall I test both behavior by creating a separate contribution for eve ?

seanlip replied:
This test is meant to check privileges, so don't overload it; it's not meant to check correct redirection.

I think it's fine to check here that Eve is on the library page. You could perhaps add the check for the dashboard page to some other protractor test that involves creating an exploration, though I'd expect there might be a test that already covers that. Please double-check, though.

thedeveloperr replied:
Done

seanlip commented on main.py:
```diff
@@ -71,7 +71,20 @@ class HomePageRedirectHandler(base.BaseHandler):
-            self.redirect(feconf.DASHBOARD_URL)
+
+            # Get all the explorations associated with the user's account
+            user_exps = user_services.get_user_contributions(self.user_id)
```
The name should be user_contributions

thedeveloperr replied:
Done

seanlip commented on main.py:
```diff
@@ -71,7 +71,20 @@ class HomePageRedirectHandler(base.BaseHandler):
+            # If the user is not learner it will have atleast
+            # one exloration associated with the user account
+            if len(user_exps.created_exploration_ids) >= 1:
+
```
You don't need all the extra blank lines here, and some of the comments are unnecessary (see above).

thedeveloperr replied:
Done

seanlip commented:
@thedeveloperr -- how's this going?

thedeveloperr commented:
@seanlip apologies for no activity recently. Due to an unfortunate reason, I was stuck in a city and had no laptop or computer facility for past 4 days. I am back now. I am currently writing automated tests for testing for editors. Will make a commit by tomorrow.

thedeveloperr commented:
@seanlip Please check latest changes

seanlip commented on core/controllers/base_test.py:
```diff
@@ -95,16 +114,72 @@ def test_requests_for_invalid_paths(self):
+            exploration_id, creator_user_id, title='Test',
+            category='Test', language_code='en')
+
+        # Since atleast one explorations have been created, going to '/' should
```
Grammar: at least one exploration has

thedeveloperr replied:
Done

seanlip commented on core/controllers/base_test.py:
```diff
@@ -95,16 +114,72 @@ def test_requests_for_invalid_paths(self):
+        self.login(self.TEST_CREATOR_EMAIL)
+        response = self.testapp.get('/')
+
+        # Must be on dashboard even though exploration is deleted because even
```
Even though the exploration is deleted, the user is still redirected to the dashboard. This is because deleted explorations are still associated with their creators.

thedeveloperr replied:
Done

seanlip commented on core/controllers/base_test.py:
```diff
@@ -95,16 +114,72 @@ def test_requests_for_invalid_paths(self):
+                'new_value': 'edited category'
+            }], 'Change title and category')
+
+        # Since user has edited one explorations created by another user
```
explorations --> exploration

thedeveloperr replied:
Done

seanlip commented on main.py:
```diff
@@ -71,7 +71,17 @@ class HomePageRedirectHandler(base.BaseHandler):
+            user_contributions = user_services.get_user_contributions(
+                self.user_id)
+
+            # 'Creator' is a user who have created or edited an exploration.
```
have --> has

thedeveloperr replied:
Done

seanlip commented on main.py:
```diff
@@ -71,7 +71,17 @@ class HomePageRedirectHandler(base.BaseHandler):
+                self.user_id)
+
+            # 'Creator' is a user who have created or edited an exploration.
+            userIsCreator = (
```
user_is_creator

thedeveloperr replied:
Done

seanlip commented:
@thedeveloperr -- please could you reply to all comments, following the instructions in step 5 [here](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change), before requesting a review? In particular, for comments that you've addressed, please reply "Done." -- this helps keep track of which comments have been addressed and which haven't. 

Make sure you don't miss any, btw -- I think you missed replying to at least one in the previous review. Thanks!

seanlip commented on core/controllers/base_test.py:
```diff
@@ -95,16 +114,72 @@ def test_requests_for_invalid_paths(self):
+            exploration_id, creator_user_id, title='Test',
+            category='Test', language_code='en')
+
+        # Since atleast one exploration has been created, going to '/' should
```
"at least" is two words.

thedeveloperr replied:
Done

seanlip commented on core/controllers/base_test.py:
```diff
@@ -95,16 +114,72 @@ def test_requests_for_invalid_paths(self):
+                'new_value': 'edited category'
+            }], 'Change title and category')
+
+        # Since user has edited one exploration created by another user
```
Add a comma at the end of this line, otherwise it's hard to tell where the clauses in the sentence end.

thedeveloperr replied:
Done

thedeveloperr replied:
I apologize for my English. It is not my first language. I didn't double check my comments in haste. I will try to do better in future.
